expirationd.lua

```diff
@@ -258,9 +258,29 @@ local function worker_loop(task)
     -- detach worker from the guardian and attach it to sched fiber
     fiber.name(string.format("worker of %q", task.name), { truncate = true })
 
+    -- https://www.tarantool.io/en/doc/latest/reference/configuration/#confval-read_only
+    local space = box.space[task.space_id]
+    local is_ro_writable = space.temporary or space.is_local
+    local is_skipped = false
     while true do
-        local replication = box.cfg.replication_source or box.cfg.replication
-        if replication == nil or #replication == 0 or task.force then
+        if not box.info.ro or is_ro_writable or task.force then
+            if is_skipped then
+                -- Wait for maximum upstream lag * 2 if we became a master.
+                -- It helps to don't hit a conflict due to parallel processing
+                -- of remaining transactions from the old master and from the
+                -- new master.
+                local max_lag = 0
+                for _, r in ipairs(box.info.replication) do
```

Comment:
Seems you should also check that r is not nil.

AFAIK there could be gaps in case if user manually clears _cluster space. See https://github.com/tarantool/metrics/pull/71/files

Reply:
Oh, it is unexpected. Thank you!

`ipairs` not always works with too big gaps, so I prefer to use `table.maxn` this purpose:
```patch
diff --git a/expirationd.lua b/expirationd.lua
index ccbd396..ccd7d12 100644
--- a/expirationd.lua
+++ b/expirationd.lua
@@ -270,8 +270,9 @@ local function worker_loop(task)
                 -- of remaining transactions from the old master and from the
                 -- new master.
                 local max_lag = 0
-                for _, r in ipairs(box.info.replication) do
-                    if r.upstream and max_lag < r.upstream.lag then
+                for i = 1, table.maxn(box.info.replication) do
+                    local r = box.info.replication[i]
+                    if r and r.upstream and max_lag < r.upstream.lag then
                         max_lag = r.upstream.lag
                     end
                 end
```